src/imitation/scripts/expert_demos.py

```diff
@@ -142,6 +172,14 @@ def callback(locals_: dict, _) -> bool:
         if policy_save_interval > 0 and step % policy_save_interval == 0:
           output_dir = os.path.join(policy_dir, f'{step:05d}')
           serialize.save_stable_model(output_dir, policy, vec_normalize)
+        if ray_tune_interval > 0 and step % ray_tune_interval == 0:
+          # TODO(shwang): Not sure if `venv` has the ground truth reward
+          # after it is wrapped in VecNormalize (does VecNormalize normalize
+          # both rewards and observations?) Check this with Adam.
+          mean_return = util.rollout.mean_return(policy, venv,
```

Comment:
Most (all?) RL algorithms report mean episode reward obtained during training, so doing evaluation periodically is unnecessary and will slow things down slightly. However, I don't have strong feelings about this. It's a little awkward (but possible) to intercept the logging entry from the RL algorithm, and you have to handle different keys used by different RL algorithms, so this approach is probably cleaner.

Reply:
Extracting rewards from a vectorized monitors seems like a pain. I could imagine wrapping the VecEnv result from `make_vec_env` in a custom `VecEnvWrapper` that knows how to access all the inner monitor logs and concatenate the dataframes from `stable_baselines...monitor.load_result`.

I'm partial to making an ad-hoc-y context that temporarily sets `venv.training = False` if `venv` is a VecNormalize (this will have to assume that we never wrap `VecNormalize` in order to work). Within this context we could call `util.rollout.mean_return(venv)` to get unnormalized rewards and also don't have to worry about this evaluation affecting the statistics tracked by `VecNormalize`.

Reply:
I'm taking this as being in response to both this and the above comment I made about `VecNormalize` clipping reward.

I don't think 2) would work: setting `training` to false just stops it updating statistics, it'll still clip reward & rescale it. I guess you could have a context that changes `self.norm_reward` and `self.clip_reward` but in any case this feels hacky. If you wanted to do that I'd just an outer wrapper that wraps an inner wrapp and lets you temporarily bypass the inner wrapper.

What's the purpose of the rollouts here? As far as I can tell it's just for logging. Do we need to log this at all, given PPO2 (and other RL algorithms) already log mean rewards during training? We can perhaps eliminate `ray.tune.track.log` entirely, or if we want it just log everything that Stable Baselines logs (see https://github.com/HumanCompatibleAI/adversarial-policies/blob/master/src/modelfree/multi/train_worker.py#L13 for an example of how to do this). This would be my favoured solution if possible.

However, it may be we actually need to know the mean reward in code. For example, if you wanted to do early stopping in Ray based on this, or if you were dissatisfied with what Stable Baselines logs. In that case a solution like (1) seems cleanest. I would not want to use `load_result`: that reads from a CSV that is still being written, which invites problems with buffering, race conditions, etc. Instead why not just look for the `episode` key in the `info` dict, which is how `Monitor` is intended to be used (and how PPO2 etc reports things)? If that doesn't work, accessing the `ep_rew` property directly still feels better than loading a CSV.

Note scripts like `eval_policy` could also run into this issue (we restore the `VecNormalize`), but here since we are not training a solution like 2) seems more palatable (only a one-off change, not messing around with contexts). It'd be nice not to have to do this though.

Reply:
Decided not to use `Monitor`. As discussed in [design docs](https://www.notion.so/chaiberkeley/PR-88-Transfer-learning-from-reward-models-Ray-a2dbae9c22014c8ebc14e6c1740dcc35#16acf51dd42145f9be0024973e6c1778) I decided to use Tensorboard+SB.

Reply:
I think I will use non-context-variant of 2) for evaluating mean episode reward in `eval_policy`. I'm leaning towards using 2) for calculating the results dictionary at the end of `expert_demos_ex` as well.

Reply:
Will address Tensorboard+SB in a new PR.

Reply:
Also will address VecNormalize in a new PR.